design-proposals/klip-63-pause-resume-persistent-queries.md

```diff
@@ -0,0 +1,62 @@
+# KLIP-63: PAUSE and RESUME for persistent queries 
+
+**Author**: Jim Hughes (jnh5y) | 
+**Release Target**: Unknown | 
+**Status**: In Discussion | 
+**Discussion**: 
+
+**tl;dr:** Adds the ability to pause the processing of a persistent query.
+           
+## Motivation and background
+
+Users may wish to pause a query for a number of reasons and then resume processing.  Presently, they would have to terminate a query and recreate it.  This approach means that any state would have to be restored; the goal of this work is to add the capability to pause processing temporarily while allowing Kafka consumers and state stores to be retained.
+
+## What is in scope
+
+Pausing and resuming persistent queries and the necessary language modifications to support this are in scope.
+
+## What is not in scope
+
+We are not considering the ability to pause and resume transient queries.  Additionally, modifying the consumers' state is out of scope.  (This would allow for processing to skip records or reprocess data.)
+
+## Value/Return
+
+End users will be able to manage query processing once this feature is implemented.  For example, this will allow for better control while designing data pipelines.  Pausing an upstream query will allow users to update a downstream query or other processing and iterate through design without processing all the data in a stream.
+
+## Public APIS
+
+SQL Grammar changes:
+Add PAUSE <queryID>, PAUSE ALL, RESUME <queryID>, RESUME ALL.
+
+## Design
+
+For `PAUSE <queryID>` and `RESUME <queryID>`, appropriate methods on the `KafkaStreams` client will be called to pause or 
+resume (respectively) the topology associated with the query.
+
+For `PAUSE ALL` and `RESUME ALL`, all persistent queries will be paused or resumed.
+
+## Test plan
```

Comment:
should we add a test case for "when 2 queries share an underlying resource (input or output topic or partition or state store or ksql server node)", pausing 1 does not inadvertently affect another ?

Reply:
I'm updating one of the lines to read:

"- Pausing a query does not impact other queries managed by the system.  Concretely, when multiple queris share a topic, pausing one of the queries will not stop the processing for the other queries.  This will be demonstrated with and without shared runtimes enabled."

I don't think we can share state stores at this point.  If I'm wrong about that or if sharing state stores becomes possible, then pausing one of the queries sharing a state store becomes interesting.

Reply:
In the new phrasing " the commands will be considered for compaction" , for my own understanding, compaction is triggered by rocksDB and not kafka streams (as I understand it).  What is the sequence of steps during a server restart wrt rocksDB that will trigger this compaction?

Reply:
The runtimes are KafkaStreams instances.  Originally, each persistent query got its own instance.  Each instance creates a certain (configurable) number of StreamThreads (which map to real threads, I believe).  

This would mean that a given large number of queries, the JVM would slow down due to the large number of threads computing for CPUs.  

Shared runtimes address this by packing multiple queries into a given KafkaStreams instance.  

With the original setup, pausing in AK Streams would look like skipping all processing.  With multiple queries in a runtime, AK should skip over the topologies which are paused.

Presently, queries/topologies in a shared runtime do not share input topics not state stores.

Reply:
Ah!  Compaction is overloaded.  ksqlDB has a "command compactor" which is used on server start to replay the state of the system.

If a group of commands would interact with each other, those are considered together.  For instance, a pair of create stream and drop stream commands may cancel each other; when that's the case, there's no need to create the resources only to remove it later.

With PAUSE and RESUME, those should cancel each other out in a similar manner.  The "interesting" corner case is what should happen to PAUSED queries on server restart.  I'm suggesting that they would be paused after the system comes up.  (To be fair, I haven't worked out all the details in my prototype, but I think that it is doable.)

Reply:
thank you very much for the explanations. They helped improve my naive understanding.